bluetooth: samples: Add sample showing basic channel sounding features

olivier-le-sage opened the pull request:
Adds samples showing basic channel sounding features.

https://builds.zephyrproject.io/zephyr/pr/79499/docs/samples/bluetooth/channel_sounding/README.html

olivier-le-sage commented on samples/bluetooth/channel_sounding/connected_cs/README.rst:
```diff
@@ -0,0 +1,25 @@
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/bluetooth/channel_sounding_test` in
```
note to self fix path

olivier-le-sage commented on samples/bluetooth/channel_sounding/cs_test/README.rst:
```diff
@@ -0,0 +1,27 @@
+Building and Running
+********************
+
+This sample can be found under :zephyr_file:`samples/bluetooth/channel_sounding_test` in
```
note to self fix path

eriksandgren commented on samples/bluetooth/channel_sounding/src/connected_cs.c:
```diff
@@ -0,0 +1,498 @@
+			.max_subevent_len = 6750,
+			.tone_antenna_config_selection =
+				BT_LE_CS_TONE_ANTENNA_CONFIGURATION_INDEX_ONE,
+			.phy = 0x01,
```
Could we use one of the enum values here? 
```
enum bt_le_cs_procedure_phy {
	BT_LE_CS_PROCEDURE_PHY_1M = BT_HCI_OP_LE_CS_PROCEDURE_PHY_1M,
	BT_LE_CS_PROCEUDRE_PHY_2M = BT_HCI_OP_LE_CS_PROCEDURE_PHY_2M,
	BT_LE_CS_PROCEDURE_PHY_CODED_S8 = BT_HCI_OP_LE_CS_PROCEDURE_PHY_CODED_S8,
	BT_LE_CS_PROCEDURE_PHY_CODED_S2 = BT_HCI_OP_LE_CS_PROCEDURE_PHY_CODED_S2,
};
```

eriksandgren commented on samples/bluetooth/channel_sounding/src/connected_cs.c:
```diff
@@ -0,0 +1,498 @@
+		if (role_selection == BT_CONN_LE_CS_ROLE_INITIATOR) {
+			k_sem_take(&sem_data_received, K_FOREVER);
+
+			estimate_distance(
```
Did you measure how much time the processing in `estimate_distance` takes for the default configuration used in the sample? Just curious since I see it uses some floating point

olivier-le-sage replied:
Using k_uptime_ticks() to record how much time is spent in this function I was able to estimate that it takes about 2.5-3ms to run estimate_distance() (excluding the  printk() calls at the end). This is with a bit less than 500 bytes of mostly mode-2 steps covering about 35-40 channels. So I suppose we can extrapolate that running this code for all channels would take ~5-6ms.

I'm not sure if this is considered fast or slow, I guess it depends on how fast the devices are moving. What do you think?

olivier-le-sage replied:
Using single-precision floats and CONFIG_FPU=y CONFIG_FPU_SHARING=y I'm able to reduce this by about 40%

eriksandgren replied:
I think 5-6 ms is good enough for this sample, thanks for checking

eriksandgren commented on samples/bluetooth/channel_sounding/src/cs_test.c:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	if (result->header.procedure_done_status == BT_CONN_LE_CS_PROCEDURE_COMPLETE ||
+	    result->header.procedure_done_status == BT_CONN_LE_CS_PROCEDURE_ABORTED) {
```
Why do we handle `BT_CONN_LE_CS_PROCEDURE_ABORTED` here but not in the `subevent_result_cb` for `connected_cs.c` ?

olivier-le-sage replied:
Since there's no real synchronization here I used this trick + the 5-second timeout on `sem_results_available` to get it to re-issue the CS Test command in the case of an abort

eriksandgren commented on samples/bluetooth/channel_sounding/src/distance_estimation.c:
```diff
@@ -0,0 +1,303 @@
+	double tof;
+	double tof_mean = 0.0;
+
+	/* Cumulative Moving Average */
```
Is there a reason to do this a as moving average and not just an average?

olivier-le-sage replied:
I think this code arrives at the same result. I wanted to avoid calculating a big sum to avoid potential overflows. But, to be honest, the range of a float is big enough that this probably doesn't matter.

The approach shown here might be more appropriate with fixed-point variables

eriksandgren commented:
Thanks for creating the samples! I managed to run them :+1:

olivier-le-sage commented on samples/bluetooth/channel_sounding/src/connected_cs.c:
```diff
@@ -0,0 +1,498 @@
+	n_ap = result->header.num_antenna_paths;
+
+	if (result->step_data_buf) {
+		if (result->step_data_buf->len <= STEP_DATA_BUF_LEN) {
```
Yes, I tried to keep things simple here in zephyr so I used a long GATT write, which only needs one call to bt_gatt_write(). To send more than this I think you have to split up the data and send it over multiple writes or indications, which seemed a bit complex.

For procedures with more than one subevent, you're right that it'll only use the last subevent's results. I agree this isn't so realistic, do you have a suggestion on how to do this better while keeping things relatively simple ?

eriksandgren replied:
Perhaps one could retrieve and concatenate the data from all subevents in the procedure, before doing the bt_gatt_write(). We would still have the limitation of 512 bytes of data per procedure, so I don't think it makes sense to implement this.

Another option would be to use the first subevent in the procedure instead of the last one. Reasons for doing that instead would be:
- The first subevent in a procedure is less likely to be "dropped". Currently, our controller will drop all subsequent subevents if there is an issue with one of them. For example if it can not be scheduled.
- The last subevent in a procedure often have less data then other subevents, for example if the procedure terminates because of reaching the channel map repetitions, reaching maximum number of steps in a procedure or reaching max_procedure_len

olivier-le-sage committed:
bluetooth: samples: Add samples showing basic channel sounding features

These samples demonstrate how to use the bluetooth 6.0 channel
sounding APIs. A basic distance estimation algorithm is included.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

ivaniushkov commented on samples/bluetooth/channel_sounding/src/distance_estimation.c:
```diff
@@ -0,0 +1,303 @@
+	bool swapped;
+	float temp;
+
+	for (uint16_t i = 0; i < len - 1; i++) {
```
can we use qsort() from standard library to avoid O(n^2) complexity?
Not sure how much time we actually spend sorting but I think it can improve performance a bit

rugeGerritsen commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,56 @@
+The connected CS sample shows how to set up regular channel sounding procedures on a connection
+between two devices.
+It can be found under :zephyr_file:`samples/bluetooth/channel_sounding/connected_cs`
+
```
Would it be possible to add a simple message sequence chart visualizing how the samples are obtained and exchanged?

olivier-le-sage replied:
added references to some MSCs in the spec. The rest is just a simple GATT write

rugeGerritsen commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,56 @@
+************
+
+* A board with BLE support
+* A controller that supports the Channel Sounding feature
```
Here we could suggest running it on a Nordic board with the SoftDevice Controller as this is the only known configuration which works

rugeGerritsen commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,56 @@
+the Zephyr tree.
+
+See :zephyr:code-sample-category:`bluetooth` samples for details.
+
```
Can we add a sections "Testing" and "Sample output" like we do for samples in NCS https://github.com/nrfconnect/sdk-nrf/blob/main/samples/bluetooth/event_trigger/README.rst? 

Then it will be easier for users of the sample to identify if something unexpected happens

rugeGerritsen reviewed (commented):
Sample code looks really nice and structured. Added some ideas on how to improve the documentation. 

I need to get a second 54L DK so that I can start running it :)

sean-madigan commented on samples/bluetooth/channel_sounding/src/distance_estimation.c:
```diff
@@ -0,0 +1,303 @@
+		if (step_data->packet_quality_aa_check !=
+			    BT_HCI_LE_CS_PACKET_QUALITY_AA_CHECK_SUCCESSFUL ||
+		    step_data->packet_rssi == BT_HCI_LE_CS_PACKET_RSSI_NOT_AVAILABLE ||
+		    step_data->tod_toa_reflector == BT_HCI_LE_CS_TIME_DIFFERENCE_NOT_AVAILABLE) {
```
`BT_HCI_LE_CS_TIME_DIFFERENCE_NOT_AVAILABLE` is out of the range of `int16_t tod_toa_reflector`

olivier-le-sage replied:
I'll change it back to uint16_t in hci_types.h

olivier-le-sage commented:
Refactored a bit to remove the console kconfigs/interaction, based on feedback about best practice. Now initiator and reflector are in separate folders/apps and the sample runs itself automatically on boot.

theob-pro commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,137 @@
+Requirements
+************
+
+* Two boards with BLE and Channel Sounding support (such as an nRF54L15 devkit from Nordic Semi)
+* A controller that supports the Channel Sounding feature (such as the SoftDevice Controller)
```
```suggestion
* Two boards with Bluetooth LE and Channel Sounding support (such as an nRF54L15 devkit from Nordic Semiconductor)
* A controller that supports the Channel Sounding feature (such as the Nordic SoftDevice Controller)
```

jhedberg commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,137 @@
+Requirements
+************
+
+* Two boards with Bluetooth LE and Channel Sounding support (such as an nRF54L15 devkit from Nordic Semiconductor)
+* A controller that supports the Channel Sounding feature (such as the Nordic SoftDevice Controller)
```
This has me wondering a little bit. Ideally we shouldn't need to have references to specific boards, rather there should be an easy way to search for upstream supported boards which support a certain Bluetooth core spec version or specific Bluetooth feature. Should you perhaps at least reference the exact board board target, or have a link to the board's upstream documentation page?

As for the controller reference, I think that's even worse, since you're now referring to something which doesn't exist upstream, or rather directing the user to a downstream proprietary tree. I'm not sure we should be doing such things in the upstream tree. A slightly better situation would be some other proprietary controller which *is* buildable upstream (using `west blobs`), but I'm not sure generic samples should be encouraging that either.

@carlescufi any thoughts here?

theob-pro replied:
> Ideally we shouldn't need to have references to specific boards, rather there should be an easy way to search for upstream supported boards which support a certain Bluetooth core spec version or specific Bluetooth feature.

I agree, that would be nice but this is outside of the scope of this PR.

> Should you perhaps at least reference the exact board board target, or have a link to the board's upstream documentation page?

I agree, here is a suggestion @olivier-le-sage:

```suggestion
* Two boards with Bluetooth LE and Channel Sounding support (such as an `nRF54L15 DK`_)
* A controller that supports the Channel Sounding feature
```

kartben commented on samples/bluetooth/channel_sounding/README.rst:
```diff
@@ -0,0 +1,137 @@
+   :relevant-api: bt_gap bluetooth
+
+   Use Channel Sounding functionality.
+
+
+Bluetooth: Channel Sounding
+###########################
```
```suggestion
```
